### PR TITLE
refactor(lsp): lift glob parsing to vim.glob

### PR DESCRIPTION
**Pull request by stevearc**
Moving the logic for using vim.lpeg to create a match pattern from a glob into `vim.lsp.util`. There are several places in the LSP spec that use globs, and it's very useful to have it as a generally-available utility.

This was pulled out of #26516 to make that PR simpler

### CODE REVIEW

**Comment by justinmk** [comment]
Should this not live on `vim.lpeg`? (That module is currently entirely "vendored", but that doesn't mean we can't add new members to it.) @neovim/core 

`vim.lsp.util` should be avoided, see https://github.com/neovim/neovim/issues/25272

**Comment by clason** [comment]
I think `vim.fs` is a better place, since this seems to be specifically a "file glob". @gpanders 

In general, moving such "local" functions to an exposed location should be accompanied -- and justified -- by code reuse in other parts of the API.

**Comment by stevearc** [comment]
Happy to move this into something more global, like `vim.fs` or `vim.glob`, but it's worth mentioning that LSP globs _do_ behave a bit differently than globs in some other contexts.

> `**` to match any number of path segments, **including none**

This means that a LSP glob of `**/*.js` will match `foo.js`, but that is not the case in bash, for example.

```bash
touch foo.js
ls **/*.js
# ls: cannot access '**/*.js': No such file or directory
```

**Comment by clason** [comment]
There's precedent for using LSP specifications outside of the LSP API, if that specification is deemed globally useful (e.g., snippet format).

**Comment by mfussenegger** [comment]
> I think vim.fs is a better place, since this seems to be specifically a "file glob".

When I wrote https://github.com/neovim/neovim/issues/23291#issuecomment-1817873088 I was a bit unsure about `vim.fs` because although it is related to file globbing, it is actually not touching the filesystem.

E.g. python has it's dedicated module: https://docs.python.org/3/library/glob.html?highlight=glob#glob.glob
But a dedicated module for a single function also feels overkill.

Just to have it mentioned - I'm not opposed to vim.fs


---


> This means that a LSP glob of **/*.js will match foo.js, but that is not the case in bash, for example.

In zsh `**/*.js` also matches `foo.js`. Afaik there is no real glob pattern standard?

**Comment by lewis6991** [comment]
Can we pull this out without adding it to the API?

**Comment by clason** [comment]
That was just a suggestion (and assuming the function could be used in `vim.fs`); a single `vim.glob` top-level function could also make sense (without making it `lpeg` specific, which seems to me an implementation detail)?

Alternatively, we could think about a generic `vim.regex` module that aims to fill the gap between (and maybe consolidate) Lua patterns and full Vim regex; this could specialize several common patterns, one of them globs. (Not 100% convinced of this myself, but if it would allow removing the distinction between `#match?` and `#lua-match?` in treesitter, it might be worth it?)

**Comment by clason** [comment]
Then I would look at `vim.regex` for a model of how to expose this (basically, use Lua's fake-OO to make a `Glob` object).

**Comment by justinmk** [comment]
> use Lua's fake-OO to make a `Glob` object

an abstraction on top of the (lpeg) abstraction?

**Comment by clason** [comment]
Beats having `parse` and `match`, which is a pattern we have nowhere else. (And the same argument applies to `vim.regex`.) If we _don't_ need this abstraction, why are we refactoring it out in the first place?

**Comment by stevearc** [comment]
This may not be a pattern that neovim currently uses, but it is somewhat similar to how python does [re.compile](https://docs.python.org/3/library/re.html#re.compile) and [re.match](https://docs.python.org/3/library/re.html#re.match)

I think it's reasonable to add a new wrapper object if we want to avoid exposing the internal lpeg object, but I would also like to keep the bare `match` method so for simple cases users don't have to first parse then match.

File: runtime/lua/vim/glob.lua [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    End = P(-1) * Cc(l.P(-1)),
+  })
+
+  return p:match(pattern) --[[@as vim.lpeg.Pattern?]]
+end
+
+---@param pattern string|vim.lpeg.Pattern The glob pattern (raw or parsed) to match.
+---@param s string The string to match against pattern.
+---@return boolean Whether or not pattern matches s.
+function M.match(pattern, s)
```
Comment (justinmk) [inline runtime/lua/vim/glob.lua:76]: why is this needed?

**Comment by justinmk** [comment]
> (And the same argument applies to `vim.regex`.) 

Can't do anything about `vim.regex` right now, also it doesn't wrap lpeg or any other thing that already has its own interface.

> If we _don't_ need this abstraction, why are we refactoring it out in the first place?

Only goal here was to put `parse_glob` somewhere other than `vim.lsp.util`.

File: runtime/lua/vim/glob.lua [hunk 2]
```diff
@@ -0,0 +1,84 @@
+local lpeg = vim.lpeg
+
+local M = {}
+
+--- Parses the raw glob into an |lpeg| pattern. LPeg patterns natively support the "this" or "that"
+--- alternative constructions described in the LSP spec that cannot be expressed in a standard Lua pattern.
+---
+---@param pattern string The raw glob pattern
+---@return vim.lpeg.Pattern? pattern An |lpeg| representation of the pattern, or nil if the pattern is invalid.
+function M.parse(pattern)
```
Comment (justinmk) [inline runtime/lua/vim/glob.lua:10]: If we name this `lpeg()` or `to_lpeg`, that leaves the door open for `to_regex` in the future, and also a `Glob` class if needed. But for now, `glob.lua` is just a namespace for glob-related things.

```suggestion
  function M.parse(pattern)
```

File: runtime/lua/vim/lsp/_watchfiles.lua [hunk 3]
```diff
@@ -1,93 +1,12 @@
...
-      + Cc(l.P(0)),
-    Literal = P(1) / l.P,
-    End = P(-1) * Cc(l.P(-1)),
-  })
-
-  return p:match(pattern) --[[@as vim.lpeg.Pattern?]]
-end
-
----@private
---- Implementation of LSP 3.17.0's pattern matching: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#pattern
```
Comment (justinmk) [inline runtime/lua/vim/lsp/_watchfiles.lua:78]: this should be mentioned in glob.lua

**Comment by clason** [comment]
I don't want to waste time debating, but I feel I'm not being understood, so let me (one last time) try to explain my point:  

* we do not need nor want a dedicated "Glob" object with a slew of fields and methods

* we do not need or want to convert "Globs" between different representations. 

I only suggested to expose the "glob matching" functionality exactly (!) the same way as we expose the "vim regex matching" functionality -- which happens to be via Lua "objects", but that is not the point here. It's just about hiding (and this not needing to document or worry about keeping compatibility for) the internal representation (lpeg).

**Comment by mfussenegger** [comment]
> If the Lpeg representation happens to be important, of course, then my point is moot.

I'm not sure how important it is, but one aspect of the lpeg object is that you can combine it with further lpeg patterns. In the watchfile module that's afaik used.

**Comment by stevearc** [comment]
Refactored to the `vim.glob.to_lpeg` form. Happy to continue massaging it as desired

**Review by mfussenegger — APPROVED** [review]

**Comment by justinmk** [comment]
> I only suggested to expose the "glob matching" functionality exactly (!) the same way as we expose the "vim regex matching" functionality -- which happens to be via Lua "objects", ... It's just about hiding (and this not needing to document or worry about keeping compatibility for) the internal representation (lpeg), which is just a means to an end.

But `regex` has no other (compiled/structured) representation that we could expose. Whereas Glob can already be represented as both Regex and Lpeg.

## Regarding "vim.glob" module

Do we need a `vim.glob` or could we expose "in situ" functions (`vim.lpeg:to_glob()`, `vim.regex:to_glob()`)?

- pro:
    - The "glob semantics" need to be defined somewhere: https://github.com/neovim/neovim/pull/26853/files#r1439774527

For comparison, I surveyed some glob libaries:

- https://github.com/isaacs/node-glob
    - ❌  Introduces filesystem traversal concepts.
        - For Nvim, `vim.fs` (not `vim.glob.Glob`) should own that.
- https://github.com/kthompson/glob
    - ❌  Introduces orthogonal  filesystem traversal concepts (`Glob.Files()`).
        - For Nvim, `vim.fs` (not `vim.glob.Glob`) should own that.
- https://code.visualstudio.com/api/references/vscode-api#GlobPattern
    - Does _not_ abstract a "compiled" glob, only stores the glob string.
